src/mlir/generator.cc

```diff
@@ -356,24 +405,101 @@ namespace mlir::verona
     }
 
     // Else, it should be an operation that we can lower natively
-    // TODO: Separate between unary, binary, ternary, etc.
-    // FIXME: Make this able to discern different types of operations.
-    if (name == "+")
+    if (isUnary(ast))
     {
+      return parsingError(
+        "Unary Operation '" + name + "' not implemented yet", getLocation(ast));
+    }
+    else if (isBinary(ast))
+    {
+      // Get both arguments
       auto arg0 = parseNode(getOperand(ast, 0).lock());
       if (auto err = arg0.takeError())
         return std::move(err);
       auto arg1 = parseNode(getOperand(ast, 1).lock());
       if (auto err = arg1.takeError())
         return std::move(err);
-      auto dialect = Identifier::get("type", &context);
-      auto type = mlir::OpaqueType::get(dialect, "ret", &context);
-      return genOperation(getLocation(ast), "verona.add", {*arg0, *arg1}, type);
+      // Get op name and type
+      using opPairTy = std::pair<llvm::StringRef, mlir::Type>;
+      opPairTy op = llvm::StringSwitch<opPairTy>(name)
+                      .Case("+", std::make_pair("verona.add", unkTy))
```

Comment:
I believe you can use initialiser lists here, instead of make pair, because the argument of `.Case` is defined by the string switch.  That will make it easier to extend this to a tuple if you need more values later.